### PR TITLE
fix(eval): rescale matrix by max magnitude in spectral_matrix_sign_transaction

### LINKED ISSUE #2379: bug(optimizer-geometry): the 1e-12 Frobenius floor launders small matrices into a numerically zero matrix sign
## Summary

`spectral_matrix_sign` divides by a fixed `1e-12` magnitude floor instead of the matrix's own Frobenius norm, so any matrix whose norm falls below that floor is **laundered into a numerically zero "valid" sign** rather than orthogonalized. The function's docstring promises the opposite:

> Frobenius normalization places every singular value in its convergence interval and preserves an exact zero for a zero matrix.

`alberta_framework/evaluation/optimizer_geometry.py:174-176`:

```python
norm = jnp.linalg.norm(value)
valid = jnp.all(jnp.isfinite(value)) & jnp.isfinite(norm)
x = value / jnp.maximum(norm, jnp.asarray(1e-12, dtype=value.dtype))
```

When `norm < 1e-12` the divisor is the floor, not the norm, so `‖x‖_F = ‖M‖_F / 1e-12 ≪ 1`. The NS5 recurrence `f(X) = 3/2 X − 1/2 XX^T X` is then applied to singular values far below its convergence interval: each step multiplies them by ≈1.5, so five steps return ≈`7.59 · σ / 1e-12` — a matrix that is numerically zero — while the transaction still reports `valid=True`.

In float32 the floor is reached earlier than the exponent range suggests, because the sum of squares underflows: for entries near `1e-20`, `jnp.linalg.norm` returns exactly `0.0` for a plainly nonzero matrix.

## Reproduction

```python
import jax
import jax.numpy as jnp
from alberta_framework.evaluation.optimizer_geometry import spectral_matrix_sign

m = jnp.array([[2.0, 1.0], [0.5, -1.0]], dtype=jnp.float32)
for scale in (1.0, 1e-13, 1e-15, 1e-20, 1e-25):
    tiny = m * jnp.asarray(scale, dtype=jnp.float32)
    print(scale, float(jnp.linalg.norm(tiny)), float(jnp.linalg.norm(spectral_matrix_sign(tiny))))
```

```
scale     norm(input)      norm(returned sign)   expected
1e+00     2.500000e+00     1.414202e+00          1.414202e+00
1e-13     2.500000e-13     1.201002e+00          1.414202e+00
1e-15     2.500000e-15     1.898316e-02          1.414202e+00
1e-20     0.000000e+00     1.898437e-07          1.414202e+00
1e-25     0.000000e+00     1.898438e-12          1.414202e+00
```

Every row is the same matrix up to a positive scalar, so every row must return the same sign. `valid` is `True` in all five.

Singular values of the returned "sign" (computed in float64):

| input scale | singular values | orthogonal? |
| --- | --- | --- |
| `1.0` | `[0.99999999, 0.99998346]` | yes |
| `1e-13` | `[0.96671583, 0.71264756]` | no |
| `1e-25` | `[1.69801409e-12, 8.49007043e-13]` | no — numerically zero |

The reverse-mode derivative is also lost: `jax.jacrev(spectral_matrix_sign)` is non-finite at `1e-25` scale and finite at unit scale.

## Consumer impact

`muon_ogd_dual_update_transaction` uses this sign as the update direction (`optimizer_geometry.py:260` and `:274`). On the documented empty-constraint path — pinned by `test_muon_ogd_empty_constraints_reduces_exactly_to_ns5` as reducing exactly to NS5 — the update vanishes with gradient scale instead of staying scale-free:

| gradient scale | `‖update‖_F` |
| --- | --- |
| `1.0` | `1.414202e+00` |
| `1e-13` | `1.201002e+00` |
| `1e-25` | `1.898438e-12` |

A scale-free orthogonalized direction is the entire point of the Muon-OGD step, so a small-magnitude momentum matrix silently produces no step rather than a unit-scale one.

The frozen streaming benchmark does not expose this: `_frozen_stream()` matrices have Frobenius norms between `1.15` and `2.59`, all far above the floor, which is why every published arm metric is unaffected.

## Why the floor is the wrong instrument

The existing `test_geometry_float32_overflow_is_invalid_not_laundered` pins the policy for the other end of the exponent range: when the input's own arithmetic overflows, the transaction must report `valid=False` rather than launder a plausible-looking result. The small end does the opposite — it returns a fabricated near-zero matrix with `valid=True`. A magnitude floor cannot distinguish "genuinely zero" from "small but well-conditioned"; only the norm can, and the norm is unusable in the input dtype at that magnitude.

## Proposed fix

Rescale by an exact power of two before normalizing, which is the pattern already used for the same hazard in `alberta_framework/utils/nexting.py`, `alberta_framework/core/oak.py`, `alberta_framework/core/state_builder.py`, and `alberta_framework/benchmarks/ipmnist_screening.py`:

```python
_, exponent = jnp.frexp(jnp.max(jnp.abs(value)))
rescaled = jnp.ldexp(value, -exponent)
rescaled_norm = jnp.linalg.norm(rescaled)
positive = rescaled_norm > 0
x = jnp.where(
    positive,
    rescaled / jnp.where(positive, rescaled_norm, jnp.ones_like(rescaled_norm)),
    jnp.zeros_like(rescaled),
)
```

The largest entry lands in `[0.5, 1)`, so the sum of squares can neither overflow nor underflow, and an exact zero is preserved by the zero test rather than by a magnitude floor. Because power-of-two scaling is exact, the quotient is bit-identical to `value / norm` for every input whose norm was already representable — verified bit-identical on 200 random float32 matrices spanning `2**-30` to `2**30`. `norm` stays in place as the overflow signal so the pinned invalid-on-overflow behaviour is unchanged.

## Environment

JAX 0.11.0, jaxlib 0.11.0, numpy 2.5.1, CPython 3.13.14, Windows 11 AMD64, `jax_enable_x64` False (default).

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2379

## Summary
In `spectral_matrix_sign_transaction` (`alberta_framework/evaluation/optimizer_geometry.py`), dividing by `jnp.maximum(norm, 1e-12)` caused any matrix whose Frobenius norm fell below `1e-12` (or whose sum of squares underflowed to zero near `1e-20`) to divide by the constant floor rather than its norm. This caused the NS5 recurrence to be evaluated on sub-convergent singular values $\ll 1$, returning a near-zero matrix with `valid=True` instead of an orthogonal matrix sign.

## Proposed Changes
1. Rescaled `value` by its maximum magnitude via `jnp.frexp` and `jnp.ldexp` before computing the norm, making initial normalization exact and immune to underflow across extreme scales down to $10^{-25}$.
2. Preserved exact reduction to zero for genuinely zero matrices where `max_mag == 0.0`.
3. Added regression tests in `tests/test_optimizer_geometry.py` across input scales from $10^{-13}$ to $10^{-25}$.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_optimizer_geometry.py` (29/29 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/evaluation/optimizer_geometry.py tests/test_optimizer_geometry.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/evaluation/optimizer_geometry.py` (clean).

### CODE REVIEW

**Review by mashingaan — COMMENTED** [review]
## Duplicate of #2380

I reviewed this PR at exact head `facaba678a9b023c0c55d9a6a630310104ca7457` against issue #2379 and the earlier open PR #2380.

- PR #2380 was opened on 2026-08-23, before #2499 on 2026-08-25.
- Both replace the fixed `1e-12` normalization floor in `spectral_matrix_sign_transaction` with exact power-of-two rescaling using `frexp` and `ldexp`.
- The extra `max_mag > 0.0` guard in #2499 does not create a distinct fix. The earlier implementation already preserves exact zero through the `rescaled_norm > 0` guard.
- #2380 carries broader regression coverage and has 56/56 successful checks on its exact head. At this review, #2499 has 55 completed successful checks and `ci-ok` is still queued.

The underlying bug is valid, but this PR materially duplicates the earlier independently verified implementation. Recommendation: close or consolidate #2499 in favor of #2380. I am not approving a duplicate.

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: openai / gpt-5
Client / agent tooling: codex
Contribution skill revision: SlopDotCash/slopdotcash@1a2017b9fbbcef4263da9dda8d0b4e1d0a7bff42:skills/contribute-to-asi
Attribution status: self-reported
— [asi-review]

**Comment by lalalune** [comment]
Superseded by merged #2380. Reviewed this matrix-sign rescaling diff: exact power-of-two normalization now replaces the fixed 1e-12 floor. The selected implementation has 17 passing regressions covering scale invariance, JIT, gradients, float16/bfloat16, zero input, and the Muon-OGD consumer, with required CI green. Keeping another equivalent normalization branch would add no useful behavior.
